ft_denoise_hfc.m

```diff
@@ -0,0 +1,240 @@
+%    (at your option) any later version.
+%
+%    FieldTrip is distributed in the hope that it will be useful,
+%    but WITHOUT ANY WARRANTY; without even the implied warranty of
+%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+%    GNU General Public License for more details.
+%
+%    You should have received a copy of the GNU General Public License
+%    along with FieldTrip. If not, see <http://www.gnu.org/licenses/>.
+%
+% $Id$
+
+% these are used by the ft_preamble/ft_postamble function and scripts
+ft_revision = '$Id$';
+ft_nargin   = nargin;
+ft_nargout  = nargout;
+
+% do the general setup of the function
+ft_defaults
+ft_preamble init
+ft_preamble debug
+ft_preamble loadvar    data
+ft_preamble provenance data
+ft_preamble trackconfig
+
+% the ft_abort variable is set to true or false in ft_preamble_init
+if ft_abort
+    % do not continue function execution in case the outputfile is present and the user indicated to keep it
+    return
+end
+
+% check if the input cfg is valid for this function
+cfg = ft_checkconfig(cfg, 'forbidden',  {'trial'}); % prevent accidental typos, see issue 1729
+
+% check the input data
+data = ft_checkdata(data, 'datatype', {'raw'});
+
+% check if the input cfg is valid for this function
+cfg = ft_checkconfig(cfg, 'forbidden',  {'trial'}); % prevent accidental typos, see issue 1729
+cfg = ft_checkconfig(cfg, 'forbidden',  {'channels'}); % prevent accidental typos, see issue 1729
+
+% Check if the data has a grad structure
+if ~isfield(data,'grad')
+    error(['Data needs a grad structure']);
+end
+
+% set the defaults
+cfg.L               = ft_getopt(cfg, 'L', 1);
+cfg.channel         = ft_getopt(cfg, 'channel', 'all', 1);
+cfg.trials          = ft_getopt(cfg, 'trials', 'all', 1);
+cfg.updatesens      = ft_getopt(cfg, 'updatesens', 'yes');
+cfg.feedback        = ft_getopt(cfg, 'feedback', 'yes');
+cfg.residualcheck   = ft_getopt(cfg, 'residualcheck', 'yes');
+cfg.residualthresh  = ft_getopt(cfg, 'residualthresh', 50);
+
+% select data based on cfg.channel and cfg.trials
+tmpcfg = keepfields(cfg, {'trials', 'showcallinfo'});
+
+% Select the correct channels using ft_channelselection
+tmpcfg.channel = cfg.channel;
+data   = ft_selectdata(tmpcfg, data);
+
+% restore the provenance information
+[cfg, data] = rollback_provenance(cfg, data);
+
+% Check match between input data chans and grad chans
+[x, ~] = ismember(data.grad.label,data.label);
+
+% Warn the user if there are chans in grad not in data
+num_mismatch = sum(x(:) == 0);
+
+if num_mismatch > 0 && istrue(cfg.feedback)
+    ft_warning(['Found ' num2str(num_mismatch) ' channels in grad structure'...
+        ' not present in cfg.channel. These channels will NOT be used for '...
+        'Harmonic Field Correction']);
+end
+
+% generate harmonic basis set
+opt = [];
+opt.li = cfg.L;
+opt.v = data.grad.coilpos(x,:);
+opt.o = data.grad.coilori(x,:);
+N = spm_opm_vslm(opt);
```

Comment:
If I understand well, the spm_opm_vslm function is from Tim's toolbox, which is hosted on github, but outside SPM. I would like to suggest that we don't include the function in fieldtrip/private (if anything, it would be supplied in the Fieldtrip distro in the folder external/OPM, to make explicit that it is code that is not our responsibility/fruit-of-our-thoughts). Alternatively, given that Tim's toolbox is a toolbox in its own right, we may want to require the user to 'install' it themselves, and make the availability of this function on the matlab-path  their responsibility. Availability of (external) toolboxes is typically checked by the utility funcition ft_hastoolbox.m 

As a service to the users, fieldtrip for instance includes (parts of various versions of) SPM in external/spm, and also some stuff from eeglab, brainstorm, and loads of other packages. Maintaining this is a bit tricky for actively developing code, so ideally we would start off by *not* including the toolbox there.